core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -665,11 +687,36 @@ private[log] class Cleaner(val id: Int,
       }
 
       override def shouldRetainRecord(batch: RecordBatch, record: Record): Boolean = {
+        var isRecordRetained: Boolean = true
         if (discardBatchRecords)
           // The batch is only retained to preserve producer sequence information; the records can be removed
-          false
+          isRecordRetained = false
         else
-          Cleaner.this.shouldRetainRecord(map, retainDeletesAndTxnMarkers, batch, record, stats)
+          isRecordRetained = Cleaner.this.shouldRetainRecord(map, retainDeletesAndTxnMarkers, batch, record, stats, currentTime = currentTime)
+        isRecordRetained
+      }
+
+      override def retrieveDeleteHorizon(batch: RecordBatch) : Long = {
+        isBatchDiscardable = shouldDiscardBatch(batch, transactionMetadata)
+
+        if (batch.deleteHorizonSet()) {
+          if (batch.deleteHorizonMs() > latestDeleteHorizon) {
+            latestDeleteHorizon = batch.deleteHorizonMs()
+          }
+          return batch.deleteHorizonMs() // means that we keep the old timestamp stored
+        }
+
+        // check that the control batch has been emptied of records
+        // if not, then we do not set a delete horizon until that is true
+        if (batch.isControlBatch() && !isBatchDiscardable) {
+          return -1L
+        }
+
+        val newDeleteHorizon: Long = time.milliseconds() + tombstoneRetentionMs
+        if (newDeleteHorizon > latestDeleteHorizon) {
+          latestDeleteHorizon = newDeleteHorizon
```

Comment:
This batch could be filtered later in MemoryRecords.filterTo(). So if we maintain latestDeleteHorizon here. It may not be accurate.